video/out/opengl/hwdec_cuda.c

```diff
@@ -0,0 +1,287 @@
+
+#include <cudaGL.h>
+
+struct priv {
+    struct mp_hwdec_ctx hwctx;
+    struct mp_image_params image_params;
+    GLuint gl_textures[2];
+    GLuint gl_pbos[2];
+    bool mapped;
+
+    CUcontext cuda_ctx;
+};
+
+static int check_cu(struct gl_hwdec *hw, CUresult err, const char *func)
+{
+    const char *err_name;
+    const char *err_string;
+
+    MP_TRACE(hw, "Calling %s\n", func);
+
+    if (err == CUDA_SUCCESS)
+        return 0;
+
+    cuGetErrorName(err, &err_name);
+    cuGetErrorString(err, &err_string);
+
+    MP_ERR(hw, "%s failed", func);
+    if (err_name && err_string)
+        MP_ERR(hw, " -> %s: %s", err_name, err_string);
+    MP_ERR(hw, "\n");
+
+    return -1;
+}
+
+#define CHECK_CU(x) check_cu(hw, (x), #x)
+
+static int cuda_create(struct gl_hwdec *hw)
+{
+    CUdevice device;
+    CUcontext cuda_ctx = NULL;
+    CUcontext dummy;
+    int ret = 0, eret = 0;
+
+    struct priv *p = talloc_zero(hw, struct priv);
+    hw->priv = p;
+
+    ret = CHECK_CU(cuInit(0));
+    if (ret < 0) {
+        goto error;
+    }
+
+    ///TODO: Make device index configurable
+    ret = CHECK_CU(cuDeviceGet(&device, 0));
+    if (ret < 0) {
+        goto error;
+    }
+
+    ret = CHECK_CU(cuCtxCreate(&cuda_ctx, CU_CTX_SCHED_BLOCKING_SYNC, device));
+    if (ret < 0) {
+        goto error;
+    }
+
+    p->cuda_ctx = cuda_ctx;
+
+    p->hwctx = (struct mp_hwdec_ctx){
+        .type = HWDEC_CUDA,
+        .ctx = cuda_ctx,
+    };
+    p->hwctx.driver_name = hw->driver->name;
+    hwdec_devices_add(hw->devs, &p->hwctx);
+
+ error:
+   eret = CHECK_CU(cuCtxPopCurrent(&dummy));
+   if (eret < 0) {
+       return eret;
+   }
+   return ret;
+}
+
+static int reinit(struct gl_hwdec *hw, struct mp_image_params *params)
+{
+    struct priv *p = hw->priv;
+    GL *gl = hw->gl;
+    CUcontext dummy;
+    int ret = 0, eret = 0;
+
+    assert(params->imgfmt == hw->driver->imgfmt);
+    p->image_params = *params;
+    params->imgfmt = IMGFMT_NV12;
+    params->hw_subfmt = 0;
+
+    ret = CHECK_CU(cuCtxPushCurrent(p->cuda_ctx));
+    if (ret < 0) {
+        return ret;
+    }
+
+    gl->GenTextures(2, p->gl_textures);
+    for (int n = 0; n < 2; n++) {
+        gl->BindTexture(GL_TEXTURE_2D, p->gl_textures[n]);
+        GLenum filter = GL_NEAREST;
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, filter);
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, filter);
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+        gl->TexImage2D(GL_TEXTURE_2D, 0, n == 0 ? GL_R8 : GL_RG8,
+                       params->w / (n + 1), params->h / (n + 1),
+                       0, n == 0 ? GL_RED : GL_RG, GL_UNSIGNED_BYTE, NULL);
+    }
+    gl->BindTexture(GL_TEXTURE_2D, 0);
+
+    gl->GenBuffers(2, p->gl_pbos);
+    for (int n = 0; n < 2; n++) {
+        gl->BindBuffer(GL_PIXEL_UNPACK_BUFFER_ARB, p->gl_pbos[n]);
```

Comment:
Other code in mpv doesn't use the `_ARB` suffix for PBOs.

Also I think you should check for desktop OpenGL in the init function, just to be sure (since this could also run with an old nvidia driver in ES2 mode if the user forces it). This can be done with `if(gl->version>=210)` or so.